internal/stats/metrics.go

```diff
@@ -1,29 +1,126 @@
 package stats
 
 import (
+	"github.com/edulinq/autograder/internal/config"
+	"github.com/edulinq/autograder/internal/log"
 	"github.com/edulinq/autograder/internal/timestamp"
 )
 
-type Metric interface {
+type MetricAttribute string
+type MetricType string
+
+// Keys for the attributes field inside of Metric and Query.
+const (
+	ASSIGNMENT_ID_KEY MetricAttribute = "assignment"
+	ANALYSIS_KEY      MetricAttribute = "analysis-type"
+	COURSE_ID_KEY     MetricAttribute = "course"
+	ENDPOINT_KEY      MetricAttribute = "endpoint"
+	LOCATOR_KEY       MetricAttribute = "locator"
+	SENDER_KEY        MetricAttribute = "sender"
+	TASK_TYPE_KEY     MetricAttribute = "task-type"
+	USER_EMAIL_KEY    MetricAttribute = "user"
+)
+
+// Values for the type field inside of Metric and Query.
+const (
+	API_REQUEST_STATS_TYPE        MetricType = "api-request-stats"
+	CODE_ANALYSIS_TIME_STATS_TYPE MetricType = "code-analysis-time-stats"
+	GRADING_TIME_STATS_TYPE       MetricType = "grading-time-stats"
+	TASK_TIME_STATS_TYPE          MetricType = "task-time-stats"
+)
```

Comment:
Since you have known values for both of these, you could use a Validate() function for both Metric and Query.
There you can do things like check for correct keys and empty values.

For a metric, it can be called before storing it.
And for a query, it can be called when received from the user.